Feature 1000 cycloneplotter

bikegeek opened the pull request:
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
- generate plots for use case using the CyclonePlotter ExtraTC config file at:

 https://github.com/dtcenter/METplus/blob/feature_1000_cycloneplotter/parm/use_cases/model_applications/tc_and_extra_tc/CyclonePlotter_fcstGFS_obsGFS_UserScript_ExtraTC.conf

to verify that the extraneous lines are absent.

- also run the use case at:
https://github.com/dtcenter/METplus/blob/feature_1000_cycloneplotter/parm/use_cases/model_applications/tc_and_extra_tc/Plotter_fcstGFS_obsGFS_ExtraTC.conf

This will produce more tracks as the original code was not correctly retrieving all the available tracks.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
- re-create the plot using the same two use cases above.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes ]**

- [x] Do these changes include sufficient testing updates? **[NA]**

- [x] Will this PR result in changes to the test suite? **[Yes ]**</br>

Different images produced than the original implementation, resulting in failed checks (2, because I added a test for the same thing early on in the testing cycle):
_

> ERROR: Some differences were found
> file not found (new output)
> A:
> B:/data/output/CyclonePlotter/metplus_final.conf
> 
> file not found (new output)
> A:
> B:/data/output/CyclonePlotter/cyclone/20150301.png
> 
> file not found (new output)
> A:
> B:/data/output/CyclonePlotter/cyclone/20150301.csv
> 
> Finished comparing directories

_

- [x] Please complete this pull request review by **[end of Week of October 12]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

bikegeek committed:
Part of Github issue #1000 METplus: added font size for legend label

bikegeek committed:
Github Issue #1000: major refactor to accommodate "sanitizing" of longitudes that could cross the International Date Line.

bikegeek committed:
Correctly returning the track dictionary in the get_points_by_track() method.

georgemccabe committed:
removed extra import of pandas

bikegeek committed:
Update use_case_groups.json

bikegeek committed:
Cleaned up unneccessary comments and other things in retrieve_data() method.

bikegeek committed:
Previous checkin was missing text due to IDE settings.

bikegeek committed:
Add correction to retrieve_data() to create directory where ASCII csv file will be written

bikegeek committed:
Clean up logic that writes the csv output file.

bikegeek committed:
Reinstate saving the png file (final plot), initial checkin had hard-coded filename and location.

bikegeek committed:
Add support for when no CYCLONE_PLOTTER_MODEL value is set and reindex the subsetted dataframe to avoid array indexing issues.

Commit:
Add configuration setting for LOOP_BY

bikegeek committed:
Added CYCLONE_PLOTTER_ANNOTATION_FONT_SIZE and modified size of the marker sizes to smaller values.

bikegeek committed:
Merge branch 'feature_1000_cycloneplotter' of https://github.com/dtcenter/METplus into feature_1000_cycloneplotter

bikegeek committed:
Updated config file with resolution setting, annotation size setting, and watermark on/off.

bikegeek committed:
Remove unneccessary debug logging message

bikegeek committed:
Github Issue #1000 added legend label fontsize and central longitude to configuration

bikegeek committed:
Github Issue #1000 fix line plot (now use the Geodesic coordinate reference system), remove unneccessary code and comments, add support to read the central longitude from the config file.

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/METplus into feature_1000_cycloneplotter

bikegeek committed:
Remove the setting for the central longitude, this should ALWAYS be 180 to center the map over the Pacific Ocean

bikegeek committed:
Change size of markers and remove setting for central longitude, this will always be set to 180 in the plotting script to center the map on the Pacific Ocean.

bikegeek committed:
Change class variable names to be more informative, add better commenting, cleaning up code, moving hard-coded values from the plotting method into initialization section.

bikegeek commented:
Now some use case tests are failing because previous plots don't match the new ones.

j-opatz commented:
I ran both use cases indicated in the PR and was successful in creating output that reflected the desired update: storm lines no longer cross the entire globe when they go across the PM. In this respect, the code updates were successful. Not only that, but there was update to the size of the legend box, so it no longer overlaps on the map area and the size of the font looks more appropriate for the image size. The only errors in the automated checks are explainable by new output, which @bikegeek noted in the PR.
However, I did want to note one thing: in an earlier test, the output showed one of the issues, the overlap of the 180 E/180 W text, had been resolved: 
![resolved_180_text](https://user-images.githubusercontent.com/59586397/137387474-2ddb31c6-0a47-486a-9c9b-bd402b108f59.png)
My current run shows that these values are again overlapping:</br>
![Screen Shot 2021-10-14 at 1 37 40 PM](https://user-images.githubusercontent.com/59586397/137386613-27b0a852-f29b-4a00-b45a-fe94ed038ed7.png)
 This is not listed in the issue so I'll approve the PR, but would it be possible to address this issue in this code update? If not, we can open a new issue for this.

bikegeek committed:
Github issue #1000, using the suggested gridlines command from https://github.com/SciTools/cartopy/issues/1401 and removing the original command of ax.gridlines(draw_labels=False, xlocs=[180, -180]) to prevent overlapping of 180E and 180W for central_longitude=180. Also fixed the sorting so the final ASCII (.csv) representation of the final data frame has been reindexed.

bikegeek committed:
Update use_case_groups.json

removed extra entry for met_tools to test the cycloneplotter wrapper earlier in the process.

bikegeek commented:
Requires cartopy 0.18 to remove the overlapping 180/-180 longitude labels when the central longitude=180.  Use the conda env cycloneplotter_carto_0_18 on 'seneca' in /home/met_test/.conda/envs/cycloneplotter_carto_0_18

bikegeek committed:
Update installation.rst

update cartopy version from 0.17 to 0.18 (the 0.18 version "fixes" the issue with overlapping 180/-180 longitude labels when the central longitude is set to 180).

j-opatz commented:
As @bikegeek indicated in the last comment, the bug of write-over text on the longitudes at the international dateline has been solved; it requires a new cartopy version to be used. I reran the 2 use cases and confirmed that this fix works as intended with no other undesired effects.
I suggest discussing the inclusion of the new cartopy version in the existing cyclone Conda environment with @georgemccabe, since I don't think we want/need a completely new conda environment to support.
